b3/parsers/iourt43.py

```diff
@@ -0,0 +1,1450 @@
+            ip = self.ip
+
+            if not pbid and self.cid:
+                fsa_info = self.console.queryClientFrozenSandAccount(self.cid)
+                self.pbid = pbid = fsa_info.get('login', None)
+
+            self.console.verbose("Auth with %r", {'name': name, 'ip': ip, 'pbid': pbid, 'guid': guid})
+
+            # FSA will be found in pbid
+            if not self.pbid:
+                # auth with cl_guid only
+                try:
+                    in_storage = self.auth_by_guid()
+                    # fix up corrupted data due to bug #162
+                    if in_storage and in_storage.pbid == 'None':
+                        in_storage.pbid = None
+                except Exception, e:
+                    self.console.error("Auth by guid failed", exc_info=e)
+                    self.authorizing = False
+                    return False
+            else:
+                # auth with FSA
+                try:
+                    in_storage = self.auth_by_pbid()
+                except Exception, e:
+                    self.console.error("Auth by FSA failed", exc_info=e)
+                    self.authorizing = False
+                    return False
+
+                if not in_storage:
+                    # fallback on auth with cl_guid only
+                    try:
+                        in_storage = self.auth_by_guid()
+                    except Exception, e:
+                        self.console.error("Auth by guid failed (when no known FSA)", exc_info=e)
+                        self.authorizing = False
+                        return False
+
+            if in_storage:
+                self.lastVisit = self.timeEdit
+                self.console.bot("Client found in the storage @%s: welcome back %s [FSA: '%s']", self.id, self.name, self.pbid)
+            else:
+                self.console.bot("Client not found in the storage %s [FSA: '%s'], create new", str(self.guid), self.pbid)
+
+            self.connections = int(self.connections) + 1
+            self.name = name
+            self.ip = ip
+            if pbid:
+                self.pbid = pbid
+            self.save()
+            self.authed = True
+
+            self.console.debug("Client authorized: %s [@%s] [GUID: '%s'] [FSA: '%s']", self.name, self.id, self.guid, self.pbid)
+
+            # check for bans
+            if self.numBans > 0:
+                ban = self.lastBan
+                if ban:
+                    self.reBan(ban)
+                    self.authorizing = False
+                    return False
+
+            self.refreshLevel()
+            self.console.queueEvent(self.console.getEvent('EVT_CLIENT_AUTH', data=self, client=self))
+            self.authorizing = False
+            return self.authed
+        else:
+            return False
+
+    def __str__(self):
+        return "Client43<@%s:%s|%s:\"%s\":%s>" % (self.id, self.guid, self.pbid, self.name, self.cid)
+
+
+class Iourt43Parser(Iourt41Parser):
```

Comment:
I like the fact you inherited from 4.1 and not 4.2. It makes the inheritance hierarchy less deep 👍